ltm/agent.py

```diff
@@ -6,32 +6,38 @@
 from dataclasses import dataclass, field
 import datetime
 from math import ceil
-import unicodedata
 from typing import Optional, Callable, Any, List, Tuple
 
 from goodai.helpers.json_helper import sanitize_and_parse_json, SimpleJSONEncoder, SimpleJSONDecoder
 from goodai.ltm.mem.base import RetrievedMemory
-from litellm import token_counter
 
 from ltm.memory.hybrid_memory import HybridMemory
 from ltm.utils.config import Config
+from ltm.extended_thinking import message_notes_and_analysis
 from model_interfaces.base_ltm_agent import Message
 
-from utils.llm import make_system_message, make_user_message, ask_llm, log_llm_call, make_assistant_message
-from utils.text import td_format
+from utils.llm import make_system_message, make_user_message, ask_llm, log_llm_call, make_assistant_message, \
+    LLMContext, count_tokens_for_model
+from utils.text import stamp_content
 from utils.ui import colour_print
 
 
+CostCallback = Callable[[float], None]
+
+
+def init_timestamp_factory():
+    return datetime.datetime.now().strftime("%F %T")
```

Comment:
What is the benefit of this local function over the definition in `__post_init__:69`?

Reply:
Just that `fields` can take it, but actually that makes it a worse option. I'll revert it.